master/internal/provisioner/aws_spot.go

```diff
@@ -0,0 +1,667 @@
+package provisioner
+
+import (
+	"encoding/base64"
+	"fmt"
+	"sort"
+	"strings"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/service/ec2"
+	"github.com/google/uuid"
+	"github.com/pkg/errors"
+
+	"github.com/determined-ai/determined/master/pkg/actor"
+)
+
+// Spot instances are created asynchronously. You create a spot request, the
+// request is validated and, if there is available capacity at the given price,
+// an instance will be created (spot request fulfilled). We use one-time spot
+// requests rather than persistent requests - if an instance is
+// shut down, the spot request will not try to automatically launch a new instance.
+// We do this so state management is slightly simpler
+// because AWS will not be doing any provisioning outside of our code that we need to account for.
+//
+// Once the spot request has been fulfilled, the request will have a pointer to the instance
+// id. If the spot request is cancelled, the instance will continue to run. The
```

Comment:
> Do we cancel requests while not terminating the associated instances

It is possible. We don't normally do it - if the request is fulfilled, we terminate the instance and that automatically cancels the spot request. However, if we think a spot request hasn't been fulfilled, we will only cancel the request. There is a case where we could delete the spot request, but (temporarily) leave the instance running:
1. query the API
2. API response says the request has not been fulfilled 
3. we tell the `scaleDecider` that it is a `pendingSpotRequest` 
4. the `scaleDecider` tells the provisioner to terminate the `pendingSpotRequest` 
5. the spot request gets fulfilled, but we have no way of knowing this 
6. we terminate the spot request and a running instance is left behind

This is the race condition mentioned in the comments, but we will clean up those instances the next time we run `listSpot`

Reply:
And I agree, but the vast majority of the effort in spot is around understanding how spot requests work. I thought it was worth documenting in this file since anyone who needs to understand or edit it in the future will need to know this info (and not all of it is documented by AWS).